src/azul/service/manifest_service.py

```diff
@@ -776,6 +809,409 @@ def manifest_config(self) -> ManifestConfig:
+        self._entities_to_relations[file_entity] = set()
+        for entity_type, objects in contents.items():
+            # FIXME: Checking the type here is working around the unexpected
+            # attribute contents.total_estimated_cells
+            #  https://github.com/DataBiosphere/azul/issues/2298
+            if isinstance(objects, list):
+                if entity_type != 'files':
+                    if len(objects) == 0:
+                        continue
+                    elif len(objects) == 1:
+                        entity = PFBEntity(name=entity_type, object=one(objects))
+                        self._entities_to_relations[entity].add(Relation.to_entity(file_entity))
+                    else:
+                        raise ValueError(f'Elasticsearch document {doc} had unexpected {entity_type!r} entities')
+
+    def entities(self) -> Iterable[PFBEntity]:
+        for entity, relations in self._entities_to_relations.items():
+            yield entity.to_json(relations)
+
+
+class PFBManifestGenerator(FileBasedManifestGenerator):
+
+    @property
+    def file_name_extension(self) -> str:
+        return 'avro'
+
+    @property
+    def content_type(self) -> str:
+        return 'application/octet-stream'
+
+    @property
+    def entity_type(self) -> str:
+        return 'files'
+
+    @property
+    def source_filter(self) -> SourceFilters:
+        """
+        We want all of the metadata because then we can use the field_types()
+        to generate the complete schema.
+        """
+        return []
+
+    def _all_docs(self) -> Iterable[JSON]:
+        for hit in self._create_request().scan():
+            doc = self._hit_to_doc(hit)
+            yield doc
+            file_ = one(doc['contents']['files'])
+            for related in file_['related_files']:
+                related_doc = copy_json(doc)
+                related_doc['contents']['files'] = [{**file_, **related}]
+                yield related_doc
+
+    def create_file(self) -> Tuple[str, Optional[str]]:
+        fd, path = mkstemp()
+        # This avro schema is the schema for the tables we create from the files
+        tables_schema = self.tables_schema(FileTransformer.field_types())
+        metadata_entity = self.metadata_entity(FileTransformer.field_types())
+        # fastavro doesn't know about our JSON type, hence the cast
+        parsed_schema = fastavro.parse_schema(cast(dict, self.pfb_schema(tables_schema)))
+
+        aggregator = PFBEntityAggregator()
+        for doc in self._all_docs():
+            aggregator.add_doc(doc)
+
+        with open(path, 'w+b') as fh:
+            fastavro.writer(fh, parsed_schema, [metadata_entity], validator=True)
+            for entity in aggregator.entities():
+                try:
+                    # FIXME: Writing the entities one at a time is less
+                    # efficient, but it allows us to see exactly which entities
+                    # fail. Do we keep it?
```

Comment:
Depends how much faster it is. Measure or present reasoning as to why one approach would be significantly faster than the other.